<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro" name="uarm">

<link name="world"/>

  <joint name="fixed" type="fixed">
    <origin xyz = "0 0 0.001 " rpy = "0 0 0"/>
    <parent link="world"/>
    <child link="Base"/>
  </joint>
<link name="Base">
    <static>true</static>
	<visual>
		<geometry>
			<mesh filename="package://uarm/urdf/pro_links/Base.STL"/>
		</geometry>
		<origin xyz = "0 0 0 " rpy = "0 0 0"/>
		<material name = "">
			<color rgba = "0.3 0.3 0.3 1" />
		</material>
    </visual>
	<inertial>
		<origin xyz="0.010476 0.000747 0.035226" rpy="0 0 0"/>
		<mass value="1.886"/>
		<inertia ixx="0.001196219" ixy="-0.000029358" ixz="0.000014859" 
				 iyy="0.001147997" iyz="0.000016274"  izz="0.001425617"/>
	</inertial>
    <collision>
		<geometry>
			<mesh filename="package://uarm/urdf/pro_links/Base.STL"/>
		</geometry>
    </collision>
</link>

<gazebo reference="Base">
    <material>Gazebo/Grey</material>
    <mu1>0.0</mu1>
    <mu2>0.0</mu2>
</gazebo>

<link name="Link1">
	<inertial>
		<origin xyz="-0.002175 0 0.029097" rpy="0 0 0"/>
		<mass value="0.2141"/>
		<inertia ixx="0.000496945" ixy="-0.000000082" ixz="-0.000002744" 
				 iyy="0.000150389" iyz="-0.000000002" izz="0.000522487"/>
	</inertial>
	<visual>
		<geometry>
			<mesh filename="package://uarm/urdf/pro_links/Link1.STL"/>
		</geometry>
		<origin xyz = "0 0 -0.0723" rpy = "0 0 0" />
		<material name = "">
			<color rgba = "0.3 0.3 0.3 1" />
		</material>
    </visual>
    <collision>
        <origin xyz = "0 0 -0.0723" rpy = "0 0 0" />
		<geometry>
			<mesh filename="package://uarm/urdf/pro_links/Link1.STL"/>
		</geometry>
    </collision>
</link>

<gazebo reference="Link1">
    <material>Gazebo/Grey</material>
    <mu1>0.0</mu1>
    <mu2>0.0</mu2>
</gazebo>

<joint name="Joint1" type="revolute">
	<axis xyz="0 0 1"/>
	<limit effort = "0.69" lower = "-1.46329" upper = "1.46329" velocity = "0.69" />
    <dynamics damping="0.1" friction="0"/>
    <parent link="Base"/>
    <child link="Link1"/>
	<origin xyz= "0 0 0.0723" rpy = " 0 0 0" />
</joint>

<transmission name="joint1_trans">
  <type>transmission_interface/SimpleTransmission</type>
  <actuator name="$joint1_motor">
    <mechanicalReduction>1</mechanicalReduction>
  </actuator>
  <joint name="Joint1">
    <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
  </joint>
</transmission>


<link name="Link2">
	<inertial>
		<origin xyz="-0.001071 0.001101 0.053193" rpy="0 0 0"/>
		<mass value="0.026"/>
		<inertia ixx="0.000061744" ixy="-0.000000025" ixz="-0.000000763" 
				 iyy="0.000062096" iyz="-0.000001053" izz="0.000002942"/>
	</inertial>
	<visual>
		<geometry>
			<mesh filename="package://uarm/urdf/pro_links/Link2.STL"/>
		</geometry>
		<origin xyz = "-0.0132 0 -0.1056" rpy = "0 0 0" />
		<material name = "">
			<color rgba = "0.3 0.3 0.3 1" />
		</material>
    </visual>
    <collision>
        <origin xyz = "-0.0132 0 -0.1056" rpy = "0 0 0" />
		<geometry>
			<mesh filename="package://uarm/urdf/pro_links/Link2.STL"/>
		</geometry>
    </collision>
</link>

<gazebo reference="Link2">
    <material>Gazebo/Grey</material>
    <mu1>0.0</mu1>
    <mu2>0.0</mu2>
</gazebo>

<joint name="Joint2" type="revolute">
	<axis xyz="0 1 0"/>
	<limit effort = "0.69" lower = "-3" upper = "3" velocity = "0.69" />
    <dynamics damping="0.03" friction="0"/>
    <parent link="Link1"/>
    <child link="Link2"/>
	<origin xyz= "0.0132 0 0.0333" rpy = " 0 0 0" />
</joint>

<transmission name="joint2_trans">
  <type>transmission_interface/SimpleTransmission</type>
  <actuator name="$joint2_motor">
    <mechanicalReduction>10</mechanicalReduction>
  </actuator>
  <joint name="Joint2">
    <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
  </joint>
</transmission>

<link name="Link3">
	<inertial>
		<origin xyz="0.101948 -0.00022 0.011917" rpy="0 0 0"/>
		<mass value="0.023"/>
		<inertia ixx="0.000001982" ixy="0.000000483" ixz="-0.000005373" 
				 iyy="0.000081264" iyz="0.000000008" izz="0.000080538"/>
	</inertial>
	<visual>
		<geometry>
			<mesh filename="package://uarm/urdf/pro_links/Link3.STL"/>
		</geometry>
		<origin xyz="-0.0132 0 -0.24767"/>
		<material name="">
			<color rgba = "0.3 0.3 0.3 1" />
		</material>
	</visual>
    <collision>
        <origin xyz="-0.0132 0 -0.24767"/>
		<geometry>
			<mesh filename="package://uarm/urdf/pro_links/Link3.STL"/>
		</geometry>
    </collision>
</link>

<gazebo reference="Link3">
    <material>Gazebo/Grey</material>
    <mu1>0.0</mu1>
    <mu2>0.0</mu2>
</gazebo>

<joint name="Joint3" type="revolute">
	<axis xyz="0 1 0"/>
	<parent link="Link2" />
	<child link="Link3" />
	<limit effort="0.69" lower="-1.45" upper="1.45" velocity = "0.69"/>
    <dynamics damping="0.03" friction="0"/>
	<origin xyz="0 0 0.14207 " rpy="0 0 0"/>
</joint>

<transmission name="joint3_trans">
  <type>transmission_interface/SimpleTransmission</type>
  <actuator name="$joint3_motor">
    <mechanicalReduction>1</mechanicalReduction>
  </actuator>
  <joint name="Joint3">
    <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
  </joint>
</transmission>

<link name="Link8">
	<inertial>
		<origin xyz="0.027288 0.001085 -0.007344" rpy="0 0 0"/>
		<mass value="0.012"/>
		<inertia ixx="0.000002184" ixy="0.000000020" ixz="-0.000000207" 
				 iyy="0.000002992" iyz="0.000000371" izz="0.000001832"/>
	</inertial>
	<visual>
		<geometry>
			<mesh filename="package://uarm/urdf/pro_links/Suction.STL"/>
		</geometry>
		<origin xyz = "-0.17201 0 -0.24651" rpy = "0 0 0" />
		<material name = "">
			<color rgba = "0.3 0.3 0.3 1" />
		</material>
    </visual>
    <collision>
        <origin xyz = "-0.17201 0 -0.24651" rpy = "0 0 0" />
		<geometry>
			<mesh filename="package://uarm/urdf/pro_links/Suction.STL"/>
		</geometry>
    </collision>
</link>

<gazebo reference="Link8">
    <material>Gazebo/Grey</material>
    <kd>1000000000000.0</kd>
    <mu1>1000.0</mu1>
    <mu2>1000.0</mu2>
    <minDepth>0.0001</minDepth>
    <maxVel>1</maxVel>
    <dampingFactor>0.1</dampingFactor>
</gazebo>

<joint name="Joint8" type="revolute">
	<axis xyz="0 1 0"/>
	<limit effort = "0.69" lower = "-3" upper = "3" velocity = "0.69" />
    <dynamics damping="0.001" friction="0"/>
    <parent link="Link3"/>
    <child link="Link8"/>
	<origin xyz= "0.15852 0 0" rpy = " 0 0 0" />
</joint>

<transmission name="joint8_trans">
  <type>transmission_interface/SimpleTransmission</type>
  <actuator name="$joint8_motor">
    <mechanicalReduction>1</mechanicalReduction>
  </actuator>
  <joint name="Joint8">
    <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
  </joint>
</transmission>

<link name="left_end_effector">
  <gravity>0</gravity>
  <!--visual>
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <geometry>
      <box size="0.015 0.015 0.015"/>
    </geometry>
    <material name="transparent">
      <color rgba="0 0 0 0"/>
    </material>
  </visual-->
  <inertial>
    <origin rpy="0 0 0" xyz="0.000000 0.000000 0.000000"/>
    <mass value="${0.0001}"/>
    <inertia ixx="${1.7e-08}" ixy="0" ixz="0" iyy="${1.7e-08}" iyz="0" izz="${1.7e-08}"/>
  </inertial>
</link>

<!--gazebo reference="left_end_effector">
    <material>Gazebo/Transparent</material>
    <mu1>0.0</mu1>
    <mu2>0.0</mu2>
</gazebo-->

<joint name="JointTCP" type="revolute">
 <axis xyz="0 0 1"/>
  <parent link="Link8" />
  <child link="left_end_effector" />
  <origin rpy="0 0 0" xyz="0.0565 0 -0.081" />
  <limit effort="0.2" velocity="0.005" lower="-6.3" upper="6.3" />
  <!-- <limit effort="0.2" velocity="0.005" lower="0" upper="6.3" /> -->
  <dynamics damping="0.00003" friction="0.00003"/>
</joint>

<transmission name="tcp_trans">
  <type>transmission_interface/SimpleTransmission</type>
  <actuator name="$tcp_motor">
    <mechanicalReduction>1</mechanicalReduction>
  </actuator>
  <joint name="JointTCP">
    <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
  </joint>
</transmission>

  <gazebo reference="depth_camera">
    <sensor name="depth_camera" type="depth">
        <pose frame="">0 0 0 0 0 0</pose>
        <camera name="__default__">
          <horizontal_fov>1.047</horizontal_fov>
          <image>
            <width>640</width>
            <height>480</height>
          </image>
          <clip>
            <near>0.1</near>
            <far>100</far>
          </clip>
          <noise>
            <type>gaussian</type>
            <mean>0.0</mean>
            <stddev>0.007</stddev>
          </noise>
        </camera>
        <always_on>1</always_on>
        <update_rate>60</update_rate>
        <visualize>1</visualize>

        <plugin name="camera_plugin" filename="libgazebo_ros_openni_kinect.so">
          <baseline>0.2</baseline>
          <alwaysOn>true</alwaysOn>
          <!-- Keep this zero, update_rate in the parent <sensor> tag
            will control the frame rate. -->
          <updateRate>0.0</updateRate>
          <cameraName>camera_depth</cameraName>
					
          <imageTopicName>/realsense_camera/color/image_raw</imageTopicName>
          <cameraInfoTopicName>/realsense_camera/color/camera_info</cameraInfoTopicName>
          <depthImageTopicName>/realsense_camera/depth/image_raw</depthImageTopicName>
          <depthImageCameraInfoTopicName>/realsense_camera/depth/camera_info</depthImageCameraInfoTopicName>
          <pointCloudTopicName>/realsense_camera/depth/points</pointCloudTopicName>
					
          <frameName>depth_camera</frameName>
          <pointCloudCutoff>0.5</pointCloudCutoff>
          <pointCloudCutoffMax>3.0</pointCloudCutoffMax>
          <distortionK1>0</distortionK1>
          <distortionK2>0</distortionK2>
          <distortionK3>0</distortionK3>
          <distortionT1>0</distortionT1>
          <distortionT2>0</distortionT2>
          <CxPrime>0</CxPrime>
          <Cx>0</Cx>
          <Cy>0</Cy>
          <focalLength>0</focalLength>
          <hackBaseline>0</hackBaseline>
        </plugin>
      </sensor>
    <material>Gazebo/Black</material>
  </gazebo>

<gazebo>
  <plugin name="gazebo_ros_control" filename="libgazebo_ros_control.so">
    <!--robotNamespace>/</robotNamespace-->
  </plugin>
  <plugin name="gazebo_ros_vacuum_gripper" filename="libgazebo_ros_vacuum_gripper.so">
    <!--robotNamespace>/uarm_set_pump</robotNamespace-->
    <bodyName>left_end_effector</bodyName>
    <topicName>pump_grasping</topicName>
  </plugin>
</gazebo>
</robot>
